0.7 Release post (#17)
GameDevTecnico/cubos-blog: content/blog/release/0-7.rst

Release 0.7
###########

:date: 2025-06-16 00:00:00
:category: Release

> :summary: TODO

Reviewer: Missing

> :cover: {static}/images/blog/meta/first-post/cover.png

Reviewer: Probably a different image, no?

New 0.7 Features
================

This blog post covers the most important changes, but you can check out the full changelog in our `repository <https://github.com/GameDevTecnico/cubos/blob/main/CHANGELOG.md>`_.

On the Engine
-------------

Overhauled ImGui Inspector with Hook Support :author_dim:`(RiscadoA)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Before 0.7, to inspect a *Cubos* value in the ImGui inspector, you used the `DataInspector` resource.
This resource provided methods to show and edit types through ImGui, but had no state whatsoever, which was a bit counterintuitive.
Additionally, it was not possible to externally override the behavior of the inspector, which made it impossible to customize the way certain types were displayed or edited.

> This meant that, for example, if the game developer wanted to show some special UI on the tools for a component of theirs, it would not be possible.

Reviewer 2: ```suggestion
This meant that, for instance, if the game developer wanted to display a custom UI for their tool component, it wouldn't be possible.
```

Reviewer: I'd use inspector instead of tools here. As it is worded, it could lead to confusion.
```suggestion
This meant that, for example, if the game developer wanted to show some special UI on the inspector for a component of theirs, it would not be possible.
```

This previous implementation was replaced by a new `ImGuiInspector` system argument. This can be used in any ImGui system. For example:

.. code-block:: c++

    cubos.system("inspect all velocities")
         .tagged(imguiTag)
         .call([](Query<Entity, Velocity&> velocities, ImGuiInspector inspector) {
            for (auto [ent, velocity] : velocities) {
                inspector.edit(velocity, "Velocity of " + std::to_string(ent));
            }
         });

This new inspector allows overriding its behavior for specific types by using the `ImGuiInspector::hook` method.

> This paved the way for us enhancing the inspector with new feature such as showing quaternions as Euler Angles, decomposing matrices into translation, rotation and scale, and more.

Reviewer 2: ```suggestion
This paved the way for us to enhance the inspector with new features such as displaying quaternions as Euler angles; breaking down matrices into translation, rotation, scale, and more.
```

Reviewer: Only thing I object to in Nuno's suggestion is the removal of the `and`. "and more" isn't part of the matrix decomposition enumeration, so the `and` before scale should remain to indicate its end.

Additionally, support was added for mask and enum types using their respective reflection traits.

Distance Physics Constraint :author_dim:`(joaomanita)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> In 0.7 we start to add physics constraints that the user can set and customize, whereas before there was only the penetration constraint used for collision solving.

Reviewer 2: ```suggestion
"In version 0.7, we began adding physics constraints that users can set and customize. Previously, only the penetration constraint was available for collision solving.
```

> For now we added only the distance constraint, which enforces a set distance between two physics bodies, or a distance range (rope-like).

Reviewer 2: ```suggestion
For now, we've added the distance constraint, which ensures a fixed distance between two physics bodies, or a distance range (similar to a rope).
```

Reviewer: In addition to Nuno's
```suggestion
For now we added only the distance constraint, which enforces either a set distance between two physics bodies or a distance range (rope-like).
```

To add this constraint:

.. code-block:: c++

    commands.relate(ent1, ent2,
                        DistanceConstraint{.isRigid = false,
                                           .minDistance = 0.0F,
                                           .maxDistance = 5.0F,
                                           .localAnchor1 = {0.0F, 0.0F, 0.0F},
                                           .localAnchor2 = {0.0F, 0.0F, 0.0F}});

    commands.relate(ent1, ent2,
                        DistanceConstraint{.isRigid = true,
                                           .fixedDistance = 5.0F,
                                           .localAnchor1 = {0.0F, 0.0F, 0.0F},
                                           .localAnchor2 = {0.0F, 0.0F, 0.0F}});

Where isRigid denotes whether it is a fixed distance or range constraint, and the local anchors refer to the point inside each body where

> the constraint is applied.

Reviewer 2: ```suggestion
Where isRigid indicates whether it’s a fixed distance or a range constraint, and the local Anchor values specify the points on each body where
the constraint is applied.
```

Reviewer 2: não mostrava as alterações quando eu tirada o enter entre a linha 63 e 64

podem tirar esse enter?

**TODO**

> - Add more types of contraints (spring, motor, etc)

Reviewer 2: ```suggestion
- Add more types of constraints (e.g., spring, motor, etc.).
```

> - Figure out how to tweak the biases in the distance constraint to control softness

Reviewer 2: ```suggestion
- Determine how to adjust the biases in the distance constraint to control softness.
```

Voxel Mesh Frustum Culling :author_dim:`(mkuritsu)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> This release we used the intersection utilities added in the previous release to perform frustum culling in the rendered voxel meshes.

Reviewer 2: ```suggestion
In this release, we used the intersection utilities from the previous release to perform frustum culling on rendered voxel meshes.
```

> Previously, every voxel mesh that was loaded into the engine would have its data sent to the GPU for rendering regardless if they were visible or not leading to wasted use of resources, this is what this feature solves.

Reviewer 2: ```suggestion
Previously, every voxel mesh loaded into the engine would have its data sent to the GPU for rendering, regardless of whether it was visible. This resulted in wasted resources, and that’s exactly what this feature addresses.
```

> With frustum culling only meshes inside the visible area of the camera are sent for drawing, the others being ignored, potentially increasing performance in scenes with a large amount of objects.

Reviewer 2: ```suggestion
With frustum culling, only meshes within the camera’s visible area are sent to the GPU for drawing, while others are ignored, potentially improving performance in scenes with a large number of objects.
```

> The frustum culling is performed automatically in any type of camera, `OrthographicCamera <https://docs.cubosengine.org/structcubos_1_1engine_1_1OrthographicCamera.html>`_ or `PerspectiveCamera <https://docs.cubosengine.org/structcubos_1_1engine_1_1PerspectiveCamera.html>`_, so it doesn't require any intervention to use its benefits. 

Reviewer 2: ```suggestion
Frustum culling is performed automatically with any type of camera—whether it's an `OrthographicCamera <https://docs.cubosengine.org/structcubos_1_1engine_1_1OrthographicCamera.html>`_ or a `PerspectiveCamera <https://docs.cubosengine.org/structcubos_1_1engine_1_1PerspectiveCamera.html>`_. This means no manual intervention is required to benefit from it.
```

> A useful option to debug as been added to the `Camera <https://docs.cubosengine.org/structcubos_1_1engine_1_1Camera.html>`_ component to be able to freeze (stop) the updates to the camera's frustum, keeping it in the last calculated state.

Reviewer 2: ```suggestion
Additionally, a debug option has been added to the `Camera <https://docs.cubosengine.org/structcubos_1_1engine_1_1Camera.html>`_ component, allowing you to freeze (stop) updates to the camera's frustum, keeping it in the last calculated state.
```

AABB and Collider Rework :author_dim:`(fallenatlas)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> With this release we started our rework of the Colliders to be more streamlined and easy to understand.

Reviewer 2: ```suggestion
In this release, we began reworking the Colliders to make them more streamlined and easier to understand.
```

Previously, we had the `Collider` component, which contained the AABB of the Collision Shape as well as a transform, to apply an offset to the AABB.

> This component was renamed to `ColliderAABB`, so it's purpose is clearer, and the transform was removed, since it was only affecting the AABB and not the collision shape itself.

Reviewer 2: ```suggestion
Previously, we had the `Collider` component, which contained the AABB of the Collision Shape as well as a transform to apply an offset to the AABB.
This component has now been renamed to `ColliderAABB` to clarify its purpose, and the transform has been removed, as it only affected the AABB and not the collision shape itself.
```

Reviewer: Alternatively,
```suggestion
Previously, we had the `Collider` component, which contained the AABB of the Collision Shape, as well as a transform, to apply an offset to the AABB.
This component was renamed to `ColliderAABB`, so it's purpose is clearer, and the transform was removed, since it was only affecting the AABB and not the collision shape itself.
```
Both configurations are correct, just choose which one sounds best to you.

> A `Collider` will now represent the owner entity of a collision shape. This is an important distinction, as we have plans to support collision shapes composed of multiple other shapes present in child entities of the owner entity, and we need an easy way to distinguish them in the hierarchy. 

Reviewer 2: ```suggestion
A `Collider` now represents the owner entity of a collision shape. This distinction is important, as we plan to support collision shapes composed of multiple shapes from child entities of the owner entity. We need an easy way to distinguish them in the hierarchy.
```

> The offset of the collision shapes will be set by its normal transform when the collision is in a child entity, otherwise, its always assumed to be centered. This functionally is not implemented yet, but should start being rolled out in the next releases.

Reviewer 2: ```suggestion
The offset of collision shapes will be determined by the entity's transform when the collision shape is part of a child entity. Otherwise, it is assumed to be centered. This functionality has not been implemented yet, but will begin rolling out in future releases.
```

> Additionally, we now add the `ColliderAABB` automatically upon adding a Collision Shape.

Reviewer 2: ```suggestion
Additionally, we now automatically add the `ColliderAABB` when a collision shape is added.
```

Voxel Shape Inertia :author_dim:`(fallenatlas)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Until this point, we had voxel shapes, with collisions between themselves and boxes. However, the inertia matrix for this shape was being calculated from the size of the voxel grid, which means that it wasn't very accurate, especially for shapes that had many empty spaces.

Reviewer 2: ```suggestion
Up until now, we had voxel shapes that collided with each other and with boxes. However, the inertia matrix for these shapes was calculated based on the size of the voxel grid, which led to inaccuracies, particularly for shapes with many empty spaces.
```

> Now, we have a custom algorithm, which takes each voxel as a point, and sums the inertia for each point, relative to the center of mass, to compute the total inertia matrix. 

Reviewer 2: ```suggestion
Now, we use a custom algorithm that treats each voxel as a point and sums the inertia for each point, relative to the center of mass, to compute the total inertia matrix.
```

> Currently, we only support uniform density, meaning all points have the mass corresponding to the total mass divided by the number of occupied voxels of the grid.

Reviewer 2: ```suggestion
Currently, we only support uniform density, meaning each point has a mass equal to the total mass divided by the number of occupied voxels in the grid.
```

You can check the implementation of the functions defined `here <https://docs.cubosengine.org/inertia_8hpp.html>`_ for more details.

> One important note is that, at this point, although we say we calculate the Inertia tensor for this shape, in reality, we do not guarantee that it is symmetric, which by its definition, it should be.

Reviewer 2: ```suggestion
One important note is that, while we calculate the inertia tensor for this shape, we do not guarantee it will be symmetric, even though it should be by definition.
```

Reviewer: Just a small correction to Nuno's. That first comma had no business being there.
```suggestion
One important note is that while we calculate the inertia tensor for this shape, we do not guarantee it will be symmetric, even though it should be by definition.
```

> The matrix is only guaranteed to be symmetric when the shape is symmetric, but in every case it is always possible to obtain such a matrix by finding the principal axis of inertia. 

Reviewer 2: ```suggestion
The matrix is guaranteed to be symmetric only when the shape itself is symmetric. However, in all cases, it’s possible to obtain a symmetric matrix by finding the principal axis of inertia.
```

> This is a slight optimization and makes the inertia more readable, however, it's not a priority.

Reviewer 2: ```suggestion
This is a minor optimization that makes the inertia more readable, though it's not a priority.
```

New Scene Format :author_dim:`(RiscadoA)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> With the introduction of relations in the ECS, and more specifically, the `ChildOf` relation, hierarchies became important in Cubos scenes.

Reviewer 2: ```suggestion
With the introduction of relations in the ECS, particularly the `ChildOf` relation, hierarchies became essential in Cubos scenes. 
```

Reviewer: I'd leave it as `important`. There's a slight change in meaning to `essential`.

> This became a problem, as the previous scene format did not have any friendly way to represent hierarchies - a scene was just a flat list of entities, with no parent-child relationships.

Reviewer 2: ```suggestion
This posed a problem, as the previous scene format didn’t have an intuitive way to represent hierarchies—scenes were simply flat lists of entities with no parent-child relationships.
```

Reviewer: Small formatting changing to Nuno's
```suggestion
This posed a problem, as the previous scene format didn’t have an intuitive way to represent hierarchies - scenes were simply flat lists of entities with no parent-child relationships.
```

> To solve this, we designed a new scene format, where each scene is a JSON file describing one entity, its components and relations, and its children, recursively.

Reviewer 2: ```suggestion
To address this, we designed a new scene format where each scene is a JSON file describing an entity, its components and relations, as well as its children, recursively.
```

It looks like this:

.. code-block:: javascript

    {
        "inherit": "ba19170e-65a0-47cd-9bd2-b4ea9447fe22",
        "game::Player": {},
        "#camera": {
            "cubos::engine::PerspectiveCamera": {
                "fovY": 90,
            }
        },
        "#gun": {
            "inherit": "6cb09eea-4156-4a75-b0ae-488aac843e05"
        }
    }

> This scene represents an entity which inherits from another scene, with the ID `ba19170e-65a0-47cd-9bd2-b4ea9447fe22`, i.e., we're modifying an existing scene.

Reviewer 2: ```suggestion
This scene represents an entity that inherits from another scene with the ID `ba19170e-65a0-47cd-9bd2-b4ea9447fe22`, meaning we're modifying an existing scene.```

> To this base entity, a `game::Player` component and two children entities, `camera` and `gun`, are added.

Reviewer 2: ```suggestion
To this base entity, a `game::Player` component and two child entities, `camera` and `gun`, are added.
```

> The `camera` entity also receives a component `cubos::engine::PerspectiveCamera`, with a field `fovY` set to 90, and the `gun` entity inherits from another scene, with the ID `6cb09eea-4156-4a75-b0ae-488aac843e05`.

Reviewer 2: ```suggestion
The `camera` entity also receives a `cubos::engine::PerspectiveCamera` component, with a field `fovY` set to 90, and the `gun` entity inherits from another scene, with the ID `6cb09eea-4156-4a75-b0ae-488aac843e05`.
```

> This format is much more flexible than the previous one as it allows us to expand easily upon existing scenes by inheriting from them, while also repreenting the hierarchies in a natural way.

Reviewer 2: ```suggestion
This format is much more flexible than the previous one, as it allows us to easily expand upon existing scenes by inheriting from them, while also representing hierarchies in a natural way.
```

> The inheritance system has been particularly useful in our recent demos, having been used, for example, to set different skins for players, by inheriting from a base player scene and overriding the models being used.

Reviewer 2: ```suggestion
The inheritance system has been especially useful in our recent demos, such as for setting different player skins by inheriting from a base player scene and overriding the models used.
```

Custom ImGui Styling :author_dim:`(RiscadoA)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> **TODO**

Reviewer: Missing

Optional Render Picker :author_dim:`(tomas7770)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In the `previous release's blog post <https://cubosengine.org/blog/release-06>`_, some of the graphics renderer's major
performance hogs were mentioned. One of them was SSAO, which had been made optional, so that it could be disabled on

> lower end devices. The other was the render picker.

Reviewer 2: ```suggestion
In the `previous release's blog post <https://cubosengine.org/blog/release-06>`_, we discussed some of the major performance bottlenecks in the graphics renderer. One of them was SSAO, which was made optional to allow disabling it on lower-end devices. The other was the render picker.
```

Though work on making the render picker possible to disable had already begun, it wasn't finished due to deadline constraints

> and some bugs that needed to be ironed out.

Reviewer 2: ```suggestion
Although work to make the render picker optional had already started, it wasn’t completed due to deadline constraints and some lingering bugs.
```

With this release, it's finally possible to remove the `RenderPicker <https://docs.cubosengine.org/structcubos_1_1engine_1_1RenderPicker.html>`_
component from a render target without breaking graphics altogether. While this prevents some features from
working - such as selecting entities with the mouse cursor - in most cases such features aren't needed,

> and the performance cost of render picking is too significant to neglect.

Reviewer 2: ```suggestion
With this release, it's now possible to remove the `RenderPicker <https://docs.cubosengine.org/structcubos_1_1engine_1_1RenderPicker.html>`_ component from a render target without breaking the graphics system. While this disables certain features, such as selecting entities with the mouse cursor, in most cases, those features aren't crucial, and the performance cost of render picking is too high to ignore.
```

On one particular machine, after disabling render picking, our Airships demo running at 1080p on integrated graphics doubled from 30 FPS to nearly 60!

> This shows that this change makes it much more feasible to run Cubos games on weaker hardware.

Reviewer 2: ```suggestion
On one particular machine, disabling render picking in our Airships demo running at 1080p on integrated graphics resulted in a jump from 30 FPS to nearly 60 FPS!
This shows how this change makes it much more feasible to run Cubos games on lower-end hardware.
```

Center Render Voxel Pivot :author_dim:`(kuukitenshi)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Previously, the pivot point for voxel rendering was at the bottom-left corner of each voxel grid. Now, it is at the center!

Reviewer 2: ```suggestion
Previously, the pivot point for voxel rendering was at the bottom-left corner of each voxel grid. Now, it’s at the center!
```

> This change makes it easier to align voxel meshes with other entities, as there's no longer a need to manually adjust offsets.

Reviewer 2: ```suggestion
This change makes aligning voxel meshes with other entities much easier, as there's no longer a need to manually adjust offsets.
```

On the Core
-----------

Math Vector Reflection Trait :author_dim:`(RiscadoA)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

One thing that had been bothering us for a while wa the lack for proper support for math types in the reflection system.

> This led to the serialization and inspector systems not being aware of them, and showing them as if they were just structs.

Reviewer 2: ```suggestion
One issue that had been bothering us for a while was the lack of proper support for math types in the reflection system. This led to the serialization and inspector systems being unaware of them, treating them as simple structs.
```

> We've thus added a new reflection trait, `VectorTrait`, which identifies types that represent math vectors, such as all of the `glm` vector types.

Reviewer 2: ```suggestion
To address this, we've added a new reflection trait, `VectorTrait`, which identifies types that represent math vectors, such as all the `glm` vector types.
```

Entity Destruction Observers :author_dim:`(kuukitenshi)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Previously, to detect when an entity was being destroyed, we had to manually check its destruction state.

> Now, we have an observable for entity destruction, which triggers automatically when an entity is destroyed.

Reviewer 2: ```suggestion
Previously, detecting when an entity was being destroyed required manually checking its destruction state. Now, we have an observable for entity destruction that triggers automatically when an entity is destroyed.
```

This makes it easier to handle such events, as we can simply register a callback that will be invoked upon destruction.

> It is especially useful for cleaning up resources or performing actions across systems without repeatedly checking if an entity is being destroyed.

Reviewer 2: ```suggestion
This makes handling such events easier, as we can now simply register a callback that’s invoked upon destruction. It’s especially useful for cleaning up resources or performing actions across systems without needing to repeatedly check if an entity is being destroyed.
```

Query Match Counting :author_dim:`(GalaxyCrush)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> When we wanted to count matches in queries, we needed to do a for loop to count them after the query was executed.

Reviewer 2: ```suggestion
Previously, to count matches in queries, we had to use a for loop to count them after the query was executed.
```

Reviewer: Just an addition to Nuno's
```suggestion
Previously, to count matches in queries we had to use a for loop to count them after the query was executed.
```

> Now we added a new `count` method to the `Query <https://docs.cubosengine.org/classcubos_1_1core_1_1ecs_1_1Query.html>`_ class, which returns the number of matches in the query, without the need for a separate loop outside the class.

Reviewer 2: ```suggestion
Now we added a new `count` method to the `Query <https://docs.cubosengine.org/classcubos_1_1core_1_1ecs_1_1Query.html>`_ class, which directly returns the number of matches in the query, eliminating the need for a separate loop outside the class.
```

Wrapper Reflection Trait :author_dim:`(SrGesus, RiscadoA)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> **TODO**

Reviewer: Missing

On the Tools
------------

Tesseratos Menu Bar :author_dim:`(jdbaracho)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Tesseratos now has a menu bar allowing users to enabled/disable each tool.

Reviewer 2: ```suggestion
Tesseratos now features a menu bar that lets users enable or disable each tool.
```

> This change gets Tesseratos a step closer to beeing a usable editor so that more people can build games using **Cubos**.

Reviewer 2: ```suggestion
This change brings Tesseratos a step closer to being a fully functional editor, making it easier for more people to build games using **Cubos**.
```

Auto-scroll Toggle on the Console :author_dim:`(R-Camacho)`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Previously, we had to scroll manually to view the latest debug messages in the console.

> A new checkbox now allows us to toggle auto-scroll. When enabled, the console will automatically scroll to show the most recent messages as they appear.

Reviewer 2: ```suggestion
Now, a new checkbox allows us toggle auto-scroll. When enabled, the console will automatically scroll to display the most recent messages as they appear.
```

.. image:: {static}/images/blog/release/0-7/console_auto-scroll.png

Next Steps
==========

> In the next release, which should be out by the end of **TODO**, we're planning to work on the following features:

Reviewer: Missing
